cpp/MPMCQueue.hpp

```diff
@@ -0,0 +1,133 @@
+#pragma once
+
+#include <atomic>
+#include <new>
+
+#include "Vector.hpp"
+
+
+namespace hsd
+{
+    namespace priv
+    {
+        template<typename T>
+        struct Slot
+        {
+            void destroy()
+            {
+                reinterpret_cast<T*>(&storage)->~T();
+            }
+
+            ~Slot()
+            {
+                destroy();
+            }
+
+            std::atomic<size_t> ticket = {0};
+
+            T storage;
+        };
+    }
+
+    /**
+     * Multiprocessor Multiconsumer Queue [MPMCQueue]
+     * Written by PyroFlareX, inspiration from github.com/rigtorp/MPMCQueue for concept.
+     * This is a "thread-safe" MPMC queue written for HackySTL, type and max capacity must be known on creation.
+     * Probably not the best optimized implementation, but it works well enough.
+     * Works as a stack-machine, FIFO
+    */
+    template<typename T>
+    class MPMCQueue
+    {
+    public:
+        MPMCQueue() = delete;
+        MPMCQueue(size_t capacity)  :   capacity(capacity)
+        {
+            if(capacity < 1)
+            {
+                throw std::invalid_argument("MPMC capacity is less than 1");
```

Comment:
We don't have exceptions

Reply:
What should I replace it with? The queue is not resizable for concurrency reasons, so anything less than one is UB as far as I am concerned.